README.md
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/HYPERTEGRITY-AG/OpenSim/all-lints)

# Open Smart City - Sim

Copyright (c) 2021 Will Freitag, Version: 1.1.2

## TL;DR
_oscsim_ is a lightweight tool to send test data to an Orion Context Broker or FROST-Server respectively. \
It can be used to send just a single message to test your installation and also to send thousands of messages in multiple threads in order to stress your installation. \
Creation of test-data w/o integrating any sensors is another aspect that can be handled with this tool. \
For this to end, not the complete APIs of both, Orion and FROST are implemented (it's not an Orion- or FROST-Client!), but the API needed to send Data (Contexts and Things/Datastreams/Observations resp.) is in place.  

A simple call of this tool like...
```commandline
$ oscsim -s http://myserver.com -p NGSI-V2 -ad dateObserved -an temperature,f,20.5
```
...will send one context to an Orion Context Broker at myserver.com with a payload defining a weather observation (date and temperature).  
And if you don't want to read this documentation at all: Remember that you can use --help at any time! :-)
```commandline
$ oscsim -h
Open Smart City-Sim, Copyright (c) 2021 Will Freitag, Version 1.1.2
usage: oscsim [-h] -s [protocol]host-name [-p {NGSI-V2,NGSI-LD,SensorThings-MQTT,SensorThings-HTTP}] [-i]
              [-a id] [-H key value] [-f id] [-e PREFIX] [-o POSTFIX] [-c] [-n num] [-m num] [-u]
              [-q milliseconds] [-l seconds] [-y name] [-an name,type,number[,max-number]] [-as name value]
              [-ad name] [-al name,lat,long[,max-lat,max-long]] [-ab name value] [-ai indent] [-r] [-v]
              [-d from to]
Tool to create some load on Orion Context Broker/FROST-Server.

optional arguments:
  -h, --help            show this help message and exit
  -s [protocol]host-name, --server [protocol]host-name
                        This host-name will be prepended by "https://", if protocol is omitted and appended with
                        "/v2/" (NGSI-V2), "/ngsi-ld/v1" (NGSI-LD) or "/v1.1/" (SensorThings) resp. depending on
                        the server-type (see -p/--protocol).
  -p {NGSI-V2,NGSI-LD,SensorThings-MQTT,SensorThings-HTTP,DIRECT-QL}, --protocol {NGSI-V2,NGSI-LD,SensorThings-MQTT,SensorThings-HTTP,DIRECT-QL}
                        Define the type of server. [Default: NGSI-V2]
  -i, --insert-always   [Only NGSI-V2, DIRECT-QL(mandatory) and NGSI-LD!] If set, the contexts will
                        always be inserted (via POST with option 'upsert') instead of trying to update first (via PATCH) and insert (via POST), if not existing (i.e. PATCH returns '404 Not Found').
  -a id, --datastream-id id
                        [Only SensorThings!] If set, this Datastream-Id will be used for ALL Observations,
                        instead of first searching for the Thing by it's name and the correct Datastream-Id
                        afterwards.
  -H key value, --header key value
                        Define a header by key and value.
  -f id, --first-id id  Define the first id to be used or the one to be used if '-c/--static-id' is set.
                        [Default: 1]
  -e PREFIX, --prefix PREFIX
                        If set, the prefix will be prepended to the generated id.
  -o POSTFIX, --postfix POSTFIX
                        If set, the postfix will be appended to the generated id.
  -c, --static-id       If set, the id will not increment (i.e. -n times -m messages will be sent with the same
                        id ['-f/--first-id' or '1' if omitted]).
  -n num, --num-threads num
                        Define, how many threads shall be used. [Default: 1]
  -m num, --messages num
                        Define, how many messages per thread shall be sent (ignored, if '-u/--unlimited' ist
                        set). [Default: 1]
  -u, --unlimited       If set, '-m/--messages' is ignored and infinite messages will be send (in '-n/--num-
                        threads' threads). Hit 'Ctrl-C' to interrupt or set '-l/--limit-time'.
  -q milliseconds, --frequency milliseconds
                        If set, limits the frequency of the messages sent to the given number (per thread!).
  -l seconds, --limit-time seconds
                        Only in conjunction with '-u/--unlimited': Stops after the given time in seconds.
  -y name, --type name  [Only NGSI-V2, DIRECT-QL(mandatory) and NGSI-LD!] If set, this type-name will be
                        used in the payload.
  -an name,type,number[,max-number], --attribute-number name,type,number[,max-number]
                        Define a number attribute used for the payload by 'name' (The name of the attribute, e.g.: temperature), 'type' (One of i [integer] or lc [linear counter] or f [floating point])and 'number' (The value to be used). If 'max-number' is set, the number written will be randomly between 'number' and 'max-number' (each including). Note: Multiple number attributes can be defined by repeating -an.
  -as name value, --attribute-string name value
                        Define a string attribute used for the payload by 'name' (The name of the attribute,
                        e.g.: instruction) and 'value' (the actual string). Note: Multiple string attributes can
                        be defined by repeating -as.
  -ad name, --attribute-date name
                        [Only NGSI-V2!] Define a DateTime attribute used for the payload by 'name' (The name of
                        the attribute, e.g.: dateObserved). Note: The current time is used as value. Multiple
                        DateTime attributes can be defined by repeating -ad.
  -al name,lat,long[,max-lat,max-long], --attribute-location name,lat,long[,max-lat,max-long]
                        [Only NGSI-V2!] Define a location attribute used for the payload by 'name' (The name of
                        the attribute, e.g.: position), 'lat' (The value for latitude) and 'long' (The value for
                        longitude). If 'max-lat' and 'max-long' are set, the location written will be randomly
                        between 'lat' and 'max-lat' and 'long' and 'max-long' resp. (each including). Note:
                        Multiple location attributes can be defined by repeating -al.
  -ab name value, --attribute-boolean name value
                        [Only NGSI-V2!] Define a boolean attribute used for the payload by 'name' (The name of
                        the attribute, e.g.: public) and 'value' (One of 'true', 'false' or 'toggle' [ie.
                        randomly switch between true and false]). Note: Multiple boolean attributes can be
                        defined by repeating -ab.
  -ai indent, --attribute-indent indent
                        Define the number of characters for indenting the created payload. [Default: 0]
  -r, --dry-run         Do a dry run only - giving the chance to review what WOULD be done incl. seeing what the
                        payload will look like.
  -v, --verbose         Generate verbose output.
  -d from to, --delete from to
                        If set, the entities within the given range (including "from" and "to") will be deleted.

Example #1:
oscsim -s my-host.com -H Authorization 'Bearer 039ea6d72a2f32227c2110bd8d78aae33acd6782' -H Fiware-service
curltest
One message is sent using id '1'.
The tenant 'curltest' will be used as 'Fiware-service' in the header of the post.

Example #2:
ocssim -s my-host.com -n 2 -m 50 ...git commit -a -m "removed mypy from linting - it has a bug reporting false positive"

100 messages will be sent (2 threads are sending 50 messages each).The id will be looped from '1' to '100'.

Example #3:
oscsim -s my-host.com -n 5 -m 100 -f 123 -c ...
500 messages will be sent (5 threads are sending 100 messages each).The id '123' (-f is first id) will be used
for all messages (-c is static id).

Example #4:
The payload that will be sent is constructed from the -y and the -aX parameters. Example:
oscsim -y WeatherObserved -an temperature,f,-20,50 -an precipitation,i,1,20 ...
will generate a payload looking like:
{
  "id":"1",
  "type":"WeatherObserved",
  "temperature": {
    "type": "Number",
    "value": -2.3
  },
  "precipitation": {
    "type": "Number",
    "value": 13
  },
}

Example #5:
oscsim -d 100 200 -s my-host.com -H Authorization 'Bearer 039ea6d72a2f32227c2110bd8d78aae33acd6782'

This will delete all IDs starting from 100 to 200 (inclusive).
```
## Why Not Simply Use "curl" or "Postman/Newman"?
While it's easy to think about a shell script, that runs curl in a loop adding data via the backend's REST-Api, one important drawback is, that if your server is TLS-secured ('https://...), curl isn't able to cache that TLS-handshake. That means, that EVERY call will do this handshake ending up in response times like one or more seconds. It's hard to generate load this way with hundreds of messages per second. Putting more than one URL in a single curl is a solution for this, but it's hard to aggregate the results and what, if there is more logic needed (e.g. Call this URL after that URL, but only if the first call gave you an 404...)  

Postman is a nice "alternative" for curl - not only because of its nice UI and tons of useful features. Beside all of that, it is able to run in a batch mode (together with Newman), AND it caches TLS-handshakes as well! But still it's hard to generate load, since Postman is resource consuming and when you run more than one instance simultaneously, you'll soon find out your test-machine is the bottleneck. 

# Supported Backends
This script is tested with:
* Orion Context Broker, 2.5.0
* Orion-LD, post-v0.7 (experimental)
* FROST-Server, 1.13.0-SNAPSHOT  

Other versions may be compatible, but we do not currently run tests against those.
  
# Supported Python Versions
Python 3.6 and 3.7 are fully supported and tested (Linux and Windows).  
This script may work on later versions of 3, but we do not currently run tests against those versions.

# Installation
## From Package
### Install
Run `pip[3] install ocssim`.
### Run Open Smart City-Sim
Simply run `oscsim`.
## From Source
### Get Repository
Clone or download repository from [Github](https://github.com/HYPERTEGRITY-AG/OpenSim).
### Prerequisites
By running: `pip[3] install -r requirements.txt [--user]` the required libraries can be installed at once.  
Alternatively, the required libraries can be installed manually.  
Following libraries need to be installed:
* **Requests**  \
Web-site: https://github.com/psf/requests \
Install with: `$ pip install requests`
* **Eclipse Paho™ MQTT Python Client** \
Web-site: https://github.com/eclipse/paho.mqtt.python \
Install with: `$ pip install paho-mqtt`
### Run Open Smart City-Sim
From the `src` directory run `python[3] -m oscsim`.
## What's that "ID"?
In Orion Context Broker (NGSI-XX), Contexts are stored as entities, and these entities are referred by their **"id"** (e.g. "urn:ngsi-v2:AirQualityObserved:RZ:Obsv4567"). Such an entity will then have some meta-data and one or more attributes. \
In FROST (SensorThings), It all starts with a "Thing" that has a **"name"** and an internal (numeric) id. Dependencies (Datastreams, Observations) are related to that internal id that is generated by FROST. \
oscsim uses a numeric "id" (starting with simply '1') that can easily be looped. 
This id is then used as the entity's "id" (Orion) and "name" (FROST) respectively.    
Please note: This numeric id can be prepended/appended with strings, letting it look more like a "real" entity/thing, if wanted.

## What is a "Message"?

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.0] - 2021-10-19
This release contains mainly 2 topics:
* **Direct Quantumleap loadtesting** - Starting with this release it is now possible send payload directly to Quantumleap.
* **Changes in Options** - This release has 2 new options. DIRECT-QL ehanges url and test load a bit and lc in number payload add linearily increasing integer value. Both require using insert-always option.

### Added
- Option DIRECT-QL for protocols.
- Option -lc for number payload.

### Deleted
- Nothing

### Changed
- Internal creation of number payload when using protocol NGSI-V2 or DIRECT-QL in conjunction with insert always.

## [1.1.2] - 2021-07-04
This release contains bugfixes only.

### Added
- Nothing

### Deleted
- Nothing

### Changed
- Fixed a bug regarding output in timed mode.

## [1.1.1] - 2021-07-03
This release is mostly about packaging and distributing.  
The package is renamed from _opensim_ to _oscsim_, since "opensim" is already taken on PyPI.
Some minor bugs were fixed.

### Added
- Nothing

### Deleted
- Nothing

### Changed
- Nothing

## [1.1.0] - 2021-05-30
This release contains mainly 2 topics:
* **NGSI-LD Support (experimental)** - Starting with this release the experimental Orion LD-server (Version: post-v0.7) can be used with Open Smart City-Sim.
* **Changes in Options** - This release has some changes in options in a way, that there is now a more generic way for creating necessary headers. If you are familiar with the current options (in 1.0.0), here is, how to "migrate" to the new options:
  * **Bearer token** - In 1.0.0 you might have used the -b [--bearer] option like: `... -b BEARER_TOKEN ...`. With this new release you will have to use `-H Authorization "Bearer INSERT_BEARER_TOKEN_HERE"` instead.
  * **Tenant** - If you used the -t [--tenant] option like: `... -t MY_TENANT ...` to set the Fiware-service you have to switch to `-H Fiware-service MY_TENANT` now.
  * **X-Gravitee-Api-Key** - If you used the -x [--x-api-key] option to set an X-Gravitee-Api-Key, please use `-H X-Gravitee-Api-Key MY_API_KEY` now.

### Added
- NGSI-LD support (experimental).
- Option -H [--header] can be used to add header.

### Deleted
- Removed options -b [--bearer], -x [--x-api-key] and -t [--tenant].

### Changed
- Nothing

## [1.0.0] - 2021-01-17
### Added
- Version- and Copyright-printout.
- CONTRIBUTION.md

### Deleted
- Nothing.

### Changed
- Nothing.


modules/arguments.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import argparse
import shutil
import textwrap as _textwrap

from modules import helper


def parse_arguments():
    class MultiLineFormatter(argparse.HelpFormatter):
        def __init__(self, prog, indent_increment=2, max_help_position=24):
            argparse.HelpFormatter.__init__(
                self,
                prog,
                indent_increment,
                max_help_position,
                width=shutil.get_terminal_size().columns,
            )

        def _fill_text(self, text, width, indent):
            paragraphs = text.split("|n")
            multi_line_text = ""
            for paragraph in paragraphs:
                formatted_paragraph = (
                    _textwrap.fill(
                        paragraph,
                        width,
                        initial_indent=indent,
                        subsequent_indent=indent,
                    )
                    + "\n"
                )
                multi_line_text = multi_line_text + formatted_paragraph
            return multi_line_text

    # noinspection PyTypeChecker
    parser = argparse.ArgumentParser(
        description="Tool to create some load on Orion Context Broker/FROST-Server.",
        formatter_class=MultiLineFormatter,
        allow_abbrev=False,
    )

    parser.epilog = (
        "Example #1:|n"
        "./opensim.py -s my-host.com -H Authorization "
        "'Bearer 039ea6d72a2f32227c2110bd8d78aae33acd6782' "
        "-H Fiware-service curltest|n"
        "One message is sent using id '1'.|n"
        "The tenant 'curltest' will be used "
        "as 'Fiware-service' in the "
        "header of the post.|n"
    )

    parser.epilog += (
        "|n|nExample #2:|n"
        "./opensim.py -s my-host.com -n 2 -m 50 ...|n"
        "100 messages will be sent (2 threads are sending 50 messages each)."
        "The id will be looped from '1' to '100'.|n"
    )

    parser.epilog += (
        "|n|nExample #3:|n"
        "./opensim.py -s my-host.com -n 5 -m 100 -f 123 -c ...|n"
        "500 messages will be sent (5 threads are sending 100 messages each)."
        "The id '123' (-f is first id) will be used for all messages (-c is static id).|n"
    )

    parser.epilog += (
        "|n|nExample #4:|n"
        "The payload that will be sent is constructed from the -y and "
        "the -aX parameters. Example:|n"
        "./opensim.py -y WeatherObserved -an temperature,f,-20,50 "
        "-an precipitation,i,1,20 ...|n"
        "will generate a payload looking like:|n"
        "{|n"
        '  "id":"1",|n'
        '  "type":"WeatherObserved",|n'
        '  "temperature": {|n'
        '    "type": "Number",|n'
        '    "value": -2.3|n'
        "  },|n"
        '  "precipitation": {|n'
        '    "type": "Number",|n'
        '    "value": 13|n'
        "  },|n"
        "}|n"
    )

    parser.epilog += (
        "|n|nExample #5:|n"
        "./opensim.py -d 100 200 -s my-host.com "
        "-H Authorization 'Bearer 039ea6d72a2f32227c2110bd8d78aae33acd6782'|n"
        "|n"
        "This will delete all IDs starting from 100 to 200 (inclusive)."
    )

    parser.add_argument(
        "-s",
        "--server",
        metavar="[protocol]host-name",
        dest="server",
        help='This host-name will be prepended by "https://", if protocol '
        "is omitted and appended "
        'with "/v2/" (NGSI-V2), "/ngsi-ld/v1" '
        '(NGSI-LD) or "/v1.1/" (SensorThings) resp. '
        "depending on the server-type (see -p/--protocol).",
        required=True,
    )

    parser.add_argument(
        "-p",
        "--protocol",
        choices=[
            helper.PROTOCOL_NGSI_V2,
            helper.PROTOCOL_NGSI_LD,
            helper.PROTOCOL_SENSOR_THINGS_MQTT,
            helper.PROTOCOL_SENSOR_THINGS_HTTP,
        ],
        default=helper.PROTOCOL_NGSI_V2,
        dest="protocol",
        help="Define the type of server. [Default: %s]" % helper.PROTOCOL_NGSI_V2,
    )

    parser.add_argument(
        "-i",
        "--insert-always",
        dest="insert_always",
        action="store_true",
        default=False,
        help="[Only NGSI-V2 and NGSI-LD!] If set, the contexts will always be inserted "
        "(via POST with option 'upsert') instead "
        "of trying to update first (via PATCH) and insert (via POST), if not "
        "existing (i.e. PATCH returns '404 Not Found').",
    )

    parser.add_argument(
        "-a",
        "--datastream-id",
        metavar="id",
        dest="datastream_id",
        help="[Only SensorThings!] If set, this Datastream-Id will be used for ALL "
        "Observations, instead of first searching for the Thing by it's name "
        "and the correct "
        "Datastream-Id afterwards.",
        type=int,
    )

    parser.add_argument(
        "-H",
        "--header",
        metavar=("key", "value"),
        dest="headers",
        action="append",
        nargs=2,
        help="Define a header by key and value.",
    )

    parser.add_argument(
        "-f",
        "--first-id",
        metavar="id",
        dest="first_id",
        help="Define the first id to be used or the one to be used "
        "if '-c/--static-id' is "
        "set. [Default: 1]",
        default=1,
        type=int,
    )

    parser.add_argument(
        "-e",
        "--prefix",
        dest="prefix",
        help="If set, the prefix will be prepended to the generated id.",
    )

    parser.add_argument(
        "-o",
        "--postfix",
        dest="postfix",
        help="If set, the postfix will be appended to the generated id.",
    )

    parser.add_argument(
        "-c",
        "--static-id",
        dest="static_id",
        action="store_true",
        default=False,
        help="If set, the id will not increment (i.e. -n times -m messages will be "
        "sent with the same id ['-f/--first-id' or '1' if omitted]).",
    )

    parser.add_argument(
        "-n",
        "--num-threads",
        metavar="num",
        dest="num_threads",
        help="Define, how many threads shall be used. [Default: 1]",
        default=1,
        type=int,
    )

    parser.add_argument(
        "-m",
        "--messages",
        metavar="num",
        dest="num_messages",
        help="Define, how many messages per thread shall be "
        "sent (ignored, if '-u/--unlimited' "
        "ist set). [Default: 1]",
        default=1,
        type=int,
    )

    parser.add_argument(
        "-u",
        "--unlimited",
        dest="unlimited",
        action="store_true",
        default=False,
        help="If set, '-m/--messages' is ignored and infinite messages "
        "will be send "
        "(in '-n/--num-threads' threads). "
        " Hit 'Ctrl-C' to interrupt or set '-l/--limit-time'.",
    )

    parser.add_argument(
        "-q",
        "--frequency",
        metavar="milliseconds",
        dest="frequency",
        help="If set, limits the frequency of the messages sent to the given "
        "number (per thread!).",
        type=int,
    )

    parser.add_argument(
        "-l",
        "--limit-time",
        metavar="seconds",
        dest="limit_time",
        help="Only in conjunction with '-u/--unlimited': Stops after the given "
        "time in seconds.",
        type=int,
    )

    parser.add_argument(
        "-y",
        "--type",
        metavar="name",
        dest="type",
        help="[Only NGSI-V2 and NGSI-LD!] If set, this type-name will be used in the payload.",
    )

    parser.add_argument(
        "-an",
        "--attribute-number",
        metavar="name,type,number[,max-number]",
        dest="numbers",
        action="append",
        nargs=1,
        help="Define a number attribute used for the payload by 'name' (The name "
        "of the "
        "attribute, e.g.: temperature), 'type' (One of i [integer] or "
        "f [floating point])"
        "and 'number' (The value to be used). If 'max-number' is set, the "
        "number written will be"
        " randomly between 'number' and 'max-number' (each including). "
        "Note: Multiple "
        "number attributes can be defined by repeating -an.",
    )

    parser.add_argument(
        "-as",
        "--attribute-string",
        metavar=("name", "value"),
        dest="strings",
        action="append",
        nargs=2,
        help="Define a string attribute used for the payload by 'name' "
        "(The name of the attribute, e.g.: instruction)"
        " and 'value' (the actual string). Note: Multiple "
        "string attributes can be defined by repeating -as.",
    )

    parser.add_argument(
        "-ad",
        "--attribute-date",
        metavar="name",
        dest="date_times",
        action="append",
        nargs=1,
        help="[Only NGSI-V2!] Define a DateTime attribute used for the payload "
        "by 'name' "
        "(The name of the attribute, "
        "e.g.: dateObserved). Note: The current time is used as value. "
        "Multiple DateTime "
        "attributes can be defined by repeating -ad.",
    )

    parser.add_argument(
        "-al",
        "--attribute-location",
        metavar="name,lat,long[,max-lat,max-long]",
        dest="locations",
        action="append",
        nargs=1,
        help="[Only NGSI-V2!] Define a location attribute used for the payload "
        "by 'name' (The name "
        "of the attribute, e.g.: position), 'lat' (The value for latitude) "
        "and 'long' "
        "(The value for longitude). If 'max-lat' and 'max-long' are set, "
        "the location "
        "written will be randomly between 'lat' and 'max-lat' and 'long' "
        "and 'max-long' "
        "resp. (each including). Note: Multiple location attributes "
        "can be defined by repeating -al.",
    )

    parser.add_argument(
        "-ab",
        "--attribute-boolean",
        metavar=("name", "value"),
        dest="booleans",
        action="append",
        nargs=2,
        help="[Only NGSI-V2!] Define a boolean attribute used for the payload "
        "by 'name' "
        "(The name of the attribute, "
        "e.g.: public) and 'value' (One of 'true', 'false' or 'toggle' "
        "[ie. randomly switch "
        "between true and false]). Note: Multiple boolean attributes can "
        "be defined "
        "by repeating -ab.",
    )

    parser.add_argument(
        "-ai",
        "--attribute-indent",
        metavar="indent",
        dest="indent",
        help="Define the number of characters for indenting the created payload. "
        "[Default: 0]",
        default=0,
        type=int,
    )

    parser.add_argument(
        "-r",
        "--dry-run",
        dest="dry_run",
        action="store_true",
        default=False,
        help="Do a dry run only - giving the chance to review what WOULD be done "
        "incl. seeing what the "
        "payload will look like.",
    )

    parser.add_argument(
        "-v",
        "--verbose",
        dest="verbose",
        action="store_true",
        default=False,
        help="Generate verbose output.",
    )

    parser.add_argument(
        "-d",
        "--delete",
        metavar=("from", "to"),
        dest="delete",
        type=int,
        nargs=2,
        help="If set, the entities within the given range "
        '(including "from" and "to") will be deleted.',
    )

    args = parser.parse_args()

    check_arguments(parser, args)

    return args


def check_arguments(parser, args):
    if args.delete is not None:
        if args.delete[0] > args.delete[1]:
            # switch the indexes
            t = args.delete[0]
            args.delete[0] = args.delete[1]
            args.delete[1] = t
        if args.unlimited:
            parser.error(
                "The switch '-u/--unlimited' cannot be used when deleting messages "
                "[-d/--delete]!"
            )
    else:
        if args.datastream_id is not None and args.datastream_id < 0:
            parser.error(
                "Please define a positive number for Datastream-Id [-a/--datastream-id]"
            )

        if args.num_threads == 0:
            parser.error(
                "Without any thread, no messages will be sent at all! "
                "[-n/--num-threads > 0]"
            )

        if args.num_messages <= 0:
            parser.error(
                "Please consider increasing the number of messages! [-m/--messages > 0]"
            )

        if args.limit_time is not None and not args.unlimited:
            parser.error(
                "Limiting the time [-l/--limit-time] is only valid with infinite "
                "run! [-u/--unlimited]"
            )

        if args.limit_time is not None and args.limit_time <= 0:
            parser.error("Please define a positive number for limit [-l/--limit-time]")

        if args.frequency is not None and args.frequency <= 0:
            parser.error(
                "Please define a positive number for frequency [-q/--frequency]"
            )

        if args.numbers is not None:
            for number in args.numbers:
                attribute_args = number[0].split(",")
                if len(attribute_args) < 3 or len(attribute_args) > 4:
                    parser.error(
                        "-an argument ['-an %s'] expects 3 or 4 comma-delimited "
                        "parameters!" % ",".join(attribute_args)
                    )
                if attribute_args[1] == "i":
                    t = 0
                    f = 0
                    try:
                        f = int(attribute_args[2])
                    except ValueError:
                        parser.error(
                            'Please check attribute "%s": "number" must be an '
                            "integer!" % attribute_args[0]
                        )
                    if len(attribute_args) == 4:
                        try:
                            t = int(attribute_args[3])
                        except ValueError:
                            parser.error(
                                'Please check attribute "%s": "max-number" must '
                                "be an integer!" % attribute_args[0]
                            )
                        if t < f:
                            parser.error(
                                'Please check attribute "%s": "max-number" must '
                                "be greater "
                                'than or equal to "number"!' % attribute_args[0]
                            )
                elif attribute_args[1] == "f":
                    t = 0.0
                    f = 0.0
                    try:
                        f = float(attribute_args[2])
                    except ValueError:
                        parser.error(
                            'Please check attribute "%s": "number" must be a '
                            "floating point number!" % attribute_args[0]
                        )
                    if len(attribute_args) == 4:
                        try:
                            t = float(attribute_args[3])
                        except ValueError:
                            parser.error(
                                'Please check attribute "%s": "max-number" must be a '
                                "floating point number!" % attribute_args[0]
                            )
                        if t < f:
                            parser.error(
                                'Please check attribute "%s": "max-number" must be '
                                "greater "
                                'than or equal to "number"!' % attribute_args[0]
                            )
                else:
                    parser.error(
                        'Please check attribute "%s": type must be one of [i | f]!'
                        % attribute_args[0]
                    )

        if args.locations is not None:
            for location in args.locations:
                attribute_args = location[0].split(",")
                if len(attribute_args) != 3 and len(attribute_args) != 5:
                    parser.error(
                        "-al argument ['-al %s'] expects 3 or 5 comma-delimited "
                        "parameters!" % ",".join(attribute_args)
                    )
                lat_from = 0.0
                lat_to = 0.0
                long_from = 0.0
                long_to = 0.0
                try:
                    lat_from = float(attribute_args[1])
                except ValueError:
                    parser.error(
                        'Please check attribute "%s": "lat" must be a floating '
                        "point number!" % attribute_args[0]
                    )
                if lat_from < -90.0 or lat_from > 90.0:
                    parser.error(
                        'Please check attribute "%s": "lat" must be in a range from '
                        "-90.0 to 90.0!" % attribute_args[0]
                    )

                try:
                    long_from = float(attribute_args[2])
                except ValueError:
                    parser.error(
                        'Please check attribute "%s": "long" must be a floating '
                        "point number!" % attribute_args[0]
                    )
                if long_from < -180.0 or long_from > 180.0:
                    parser.error(
                        'Please check attribute "%s": "long" must be in a range from '
                        "-180.0 to 180.0!" % attribute_args[0]
                    )

                if len(attribute_args) == 5:
                    try:
                        lat_to = float(attribute_args[3])
                    except ValueError:
                        parser.error(
                            'Please check attribute "%s": "max-lat" must be a '
                            "floating "
                            "point number!" % attribute_args[0]
                        )

                    if lat_to < -90.0 or lat_to > 90.0:
                        parser.error(
                            'Please check attribute "%s": "max-lat" must be in a '
                            "range "
                            "from -90.0 to 90.0!" % attribute_args[0]
                        )

                    try:
                        long_to = float(attribute_args[4])
                    except ValueError:
                        parser.error(
                            'Please check attribute "%s": "max-long" must be a '
                            "floating point number!" % attribute_args[0]
                        )
                    if long_to < -180.0 or long_to > 180.0:
                        parser.error(
                            'Please check attribute "%s": "max-long" must be in a '
                            "range "
                            "from -180.0 to 180.0!" % attribute_args[0]
                        )

                    if lat_to < lat_from:
                        parser.error(
                            'Please check attribute "%s": "max-lat" must be greater '
                            "than "
                            'or equal to "lat"!' % attribute_args[0]
                        )

                    if long_to < long_from:
                        parser.error(
                            'Please check attribute "%s": "max-long" must be greater '
                            "than "
                            'or equal to "long"!' % attribute_args[0]
                        )

        if args.booleans is not None:
            for boolean in args.booleans:
                if (
                    boolean[1] != "true"
                    and boolean[1] != "false"
                    and boolean[1] != "toggle"
                ):
                    parser.error(
                        'Please check attribute "%s": "value" must be one '
                        'of "true", '
                        '"false" or "toggle"!' % boolean[0]
                    )

        if not args.delete:
            # check scheme?
            if args.insert_always and (
                args.protocol != helper.PROTOCOL_NGSI_V2
                and args.protocol != helper.PROTOCOL_NGSI_LD
            ):
                parser.error(
                    "Insert always scheme [-i/--insert-always] is only valid "
                    "for NGSI-V2 and NGSI-LD!"
                )

            # is there any payload?
            if (
                args.protocol == helper.PROTOCOL_NGSI_V2
                or args.protocol != helper.PROTOCOL_NGSI_LD
            ):
                if (
                    ((args.numbers is None) or (len(args.numbers) == 0))
                    and ((args.booleans is None) or (len(args.booleans) == 0))
                    and ((args.locations is None) or (len(args.locations) == 0))
                    and ((args.strings is None) or (len(args.strings) == 0))
                    and ((args.date_times is None) or (len(args.date_times) == 0))
                ):
                    parser.error("Please define any payload!")
            else:
                if ((args.numbers is None) or (len(args.numbers) == 0)) and (
                    (args.strings is None) or (len(args.strings) == 0)
                ):
                    parser.error("Please define any payload!")

    return True
